forte/dmrg/block2_dmrg_solver.cc

```diff
@@ -0,0 +1,735 @@
+                        tei_shape, tei_ba_strides, integral_cutoff, 0.5, actv_irreps);
+        r->add_sum_term(as_ints_->tei_bb_vector().data(), as_ints_->tei_bb_vector().size(),
+                        tei_shape, tei_strides, integral_cutoff, 0.25, actv_irreps);
+        r->add_sum_term(as_ints_->oei_a_vector().data(), as_ints_->oei_a_vector().size(), oei_shape,
+                        oei_strides, integral_cutoff, 1.0, actv_irreps);
+        r->add_sum_term(as_ints_->oei_b_vector().data(), as_ints_->oei_b_vector().size(), oei_shape,
+                        oei_strides, integral_cutoff, 1.0, actv_irreps);
+    }
+
+    r = r->adjust_order();
+
+    // dmrg sweep settings
+    double e_conv = dmrg_options_->get_double("BLOCK2_SWEEP_ENERGY_CONV");
+    double mps_cutoff = dmrg_options_->get_double("BLOCK2_CUTOFF");
+    int n_total_sweeps = dmrg_options_->get_int("BLOCK2_N_TOTAL_SWEEPS");
+    std::vector<int> sweep_n_sweeps = dmrg_options_->get_int_list("BLOCK2_SWEEP_N_SWEEPS");
+    std::vector<int> sweep_bond_dims = dmrg_options_->get_int_list("BLOCK2_SWEEP_BOND_DIMS");
+    std::vector<double> sweep_noises = dmrg_options_->get_double_list("BLOCK2_SWEEP_NOISES");
+    std::vector<double> sweep_davidson_tols =
+        dmrg_options_->get_double_list("BLOCK2_SWEEP_DAVIDSON_TOLS");
+    int dmrg_verbose = dmrg_options_->get_int("BLOCK2_VERBOSE");
+
+    size_t n_schedule =
+        max(sweep_n_sweeps.size(),
+            max(sweep_bond_dims.size(), max(sweep_noises.size(), sweep_davidson_tols.size())));
+    while (sweep_davidson_tols.size() < n_schedule)
+        sweep_davidson_tols.push_back(e_conv == 0.0 ? 1E-9 : e_conv * 10);
+    while (sweep_noises.size() < n_schedule)
+        if (sweep_noises.size() == n_schedule - 1)
+            sweep_noises.push_back(0.0);
+        else
+            sweep_noises.push_back(sweep_davidson_tols[sweep_noises.size()] * 10);
+    while (sweep_bond_dims.size() < n_schedule)
+        if (sweep_bond_dims.size() == 0)
+            sweep_bond_dims.push_back(500);
+        else
+            sweep_bond_dims.push_back(sweep_bond_dims.back());
+    while (sweep_n_sweeps.size() < n_schedule)
+        if (n_schedule == 1)
+            sweep_n_sweeps.push_back(n_total_sweeps == 0 ? 10 : n_total_sweeps);
+        else
+            sweep_n_sweeps.push_back(6);
+
+    // expand sweep schedule
+    int n_total_cnt = 0;
+    std::vector<int> bond_dims;
+    std::vector<double> noises, davidson_tols;
+    for (size_t i = 0; i < n_schedule; i++) {
+        n_total_cnt += sweep_n_sweeps[i];
+        for (int j = 0; j < sweep_n_sweeps[i]; j++) {
+            bond_dims.push_back(sweep_bond_dims[i]);
+            noises.push_back(sweep_noises[i]);
+            davidson_tols.push_back(sweep_davidson_tols[i]);
+        }
+    }
+
+    if (n_total_sweeps == 0)
+        n_total_sweeps = n_total_cnt;
+
+    bool read_initial_guess = options_->get_bool("READ_ACTIVE_WFN_GUESS");
```

Comment:
I wonder is there a `dump_wfn` implementation available? It seems to me that every DMRG part of the two-step MCSCF procedure always starts from a random initial guess, which leads to oscillating energies even when the orbitals are well converged.

Reply:
The MPS is automatically stored in scratch storage so there is no need to explicitly dump the MPS. One only needs to set this ``READ_ACTIVE_WFN_GUESS`` option (when it is close to convergence) to reuse the MPS from a previous MCSCF step.